src/actix/helpers.rs

```diff
@@ -70,3 +72,132 @@ where
+}
+
+async fn time_impl<T, Fut>(future: Fut) -> impl actix_web::Responder
+where
+    Fut: Future<Output = HttpResult<Option<T>>>,
+    T: serde::Serialize,
+{
+    let instant = Instant::now();
+    let result = future.await;
+    let time = instant.elapsed().as_secs_f64();
+
+    let (status_code, response) = match result {
+        Ok(result) => {
+            let (status_code, status) = if result.is_some() {
+                (http::StatusCode::OK, ApiStatus::Ok)
+            } else {
+                (http::StatusCode::ACCEPTED, ApiStatus::Accepted)
+            };
+
+            let response = ApiResponse {
+                result,
+                status,
+                time,
+            };
+
+            (status_code, response)
+        }
+
+        Err(error) => {
+            let response = ApiResponse {
+                result: None,
+                status: ApiStatus::Error(error.to_string()),
+                time,
+            };
+
+            (error.status_code(), response)
+        }
+    };
+
+    HttpResponse::build(status_code).json(response)
+}
+
+pub type HttpResult<T, E = HttpError> = Result<T, E>;
+
+#[derive(Clone, Debug, thiserror::Error)]
+#[error("{description}")]
+pub struct HttpError {
+    status_code: http::StatusCode,
+    description: String,
+}
+
+impl HttpError {
+    pub fn new(status_code: http::StatusCode, description: impl Into<String>) -> Self {
+        Self {
+            status_code,
+            description: description.into(),
+        }
+    }
+
+    pub fn status_code(&self) -> http::StatusCode {
+        self.status_code
+    }
+}
+
+impl actix_web::ResponseError for HttpError {
+    fn status_code(&self) -> http::StatusCode {
+        self.status_code
+    }
+}
+
+impl From<StorageError> for HttpError {
+    fn from(err: StorageError) -> Self {
+        let (status_code, description) = match err {
+            StorageError::BadInput { description } => (http::StatusCode::BAD_REQUEST, description),
+            StorageError::NotFound { description } => (http::StatusCode::NOT_FOUND, description),
+            StorageError::ServiceError { description, .. } => {
+                (http::StatusCode::INTERNAL_SERVER_ERROR, description)
+            }
+            StorageError::BadRequest { description } => {
+                (http::StatusCode::BAD_REQUEST, description)
+            }
+            StorageError::Locked { description } => (http::StatusCode::FORBIDDEN, description),
+            StorageError::Timeout { description } => {
+                (http::StatusCode::REQUEST_TIMEOUT, description)
+            }
+        };
+
+        Self {
+            status_code,
+            description,
+        }
+    }
+}
+
+impl From<CollectionError> for HttpError {
+    fn from(err: CollectionError) -> Self {
+        StorageError::from(err).into()
+    }
+}
+
+impl From<io::Error> for HttpError {
+    fn from(err: io::Error) -> Self {
+        StorageError::from(err).into() // TODO: Is this good enough?.. 🤔
```

Comment:
Yes :+1:

Reply:
Left the `TODO` for further consideration.